### PR TITLE
ocpdeploy user story

### PR DESCRIPTION
**Pull request by vsevel**
Here is a feedback to the community on our use of Quarkus. Please consider it as a blog post if you feel like it can inspiring for others. I am open for comments.

### CODE REVIEW

**Comment by maxandersen** [comment]
I like the blog @vsevel, but its a bit ambgious what/who the "we" refers to. Is there a way we can put a name on the "we" in the blog ?

File: _posts/2021-06-08-quarkus-deploys-apps-on-openshift.adoc [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+All was well... until we started deploying multiple tenants at the same time, and/or different applications. We had sized the _ocpdeploy_ container to `200 millicores` and `280 Mb` of RAM. For some applications we have around 30 tenants, that we all deploy in parallel. This meant 30 pods, which accounted for `6 cores` and `8 Gb` of RAM. This seems livable, but deployments tend to be done after business hours, between 7 and 8pm for many. And we started being afraid of the impact _ocpdeploy_ itself would have if we were running multiple deployments in parallel, on top of our SpringBoot or WildFly applications, which take their toll on the cluster at startup.
+
+== Enter the Quarkus Universe
+
+So less than a year ago we decided to launch a Quarkus POC on _ocpdeploy_, the goal being to deploy it as a GraalVM executable.
+The main challenges we ran into were related to the lack of support for GraalVM in some of the libraries we were using: FreeMarker, SMBJ, the Fabric8 OpenShift client (when instantiating OpenShift specific CRDs), and a home made java agent that ships logs to a proprietary centralized logging system.
```
Comment (maxandersen) [inline _posts/2021-06-08-quarkus-deploys-apps-on-openshift.adoc:46]: spell out "Server Message Block in Java" the first time possibly link to it?

File: _posts/2021-06-08-quarkus-deploys-apps-on-openshift.adoc [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+For FreeMarker we pushed a new extension to the Quarkiverse, with code largely inspired from work by _ppalaga_ and _carlosthe19916_.
+For SMBJ we created an extension in our internal Quarkiverse. And for our logging client, I was able to draw some inspiration from the `Quarkus-logging-gelf` extension, and created an additional internal extension.
+
+Getting a library to run in native was facilitated by the very good documentation that comes from the Quarkus project and support in the framework (e.g. _processors_ and _recorders_): https://Quarkus.io/guides/writing-native-applications-tips#native-in-extension[Supporting native in a Quarkus extension]
+
+When things got a little hairy (SMBJ was trickier than we thought), we got some help from the https://medium.com/GraalVM/introducingthe-tracing-agent-simplifying-GraalVM-native-image-configuration-c3b56c486271[GraalVM tracing agent].
+
+Here is the list of extensions we are currently using: `cdi, config-yaml, FreeMarker, hibernate-validator, kubernetes-client, openshift-client, rest-client, restclient-jackson, vault` (plus internal extensions for SMBJ and our internal centralized logging system).
```
Comment (maxandersen) [inline _posts/2021-06-08-quarkus-deploys-apps-on-openshift.adoc:57]:
```suggestion
Here is the list of extensions we are currently using: `cdi, config-yaml, freemarker, hibernate-validator, kubernetes-client, openshift-client, rest-client, restclient-jackson, vault` (plus internal extensions for SMBJ and our internal centralized logging system).
```

File: _posts/2021-06-08-quarkus-deploys-apps-on-openshift.adoc (hunk 1 above)
Comment (vsevel) [reply _posts/2021-06-08-quarkus-deploys-apps-on-openshift.adoc:46]: I do not understand your comment. what is "Server Message Block in Java"?

File: _posts/2021-06-08-quarkus-deploys-apps-on-openshift.adoc (hunk 2 above)
Comment (vsevel) [reply _posts/2021-06-08-quarkus-deploys-apps-on-openshift.adoc:57]: I actually respected the case of the product as stated in https://freemarker.apache.org/ 

> What is Apache FreeMarker™?

**Comment by vsevel** [comment]
> I like the blog @vsevel, but its a bit ambgious what/who the "we" refers to. Is there a way we can put a name on the "we" in the blog ?

hi @maxandersen. thanks for your comments.
the original idea was to provide a feedback to the community. @gsmet suggested that it might be a good fit for a blog post. however, corporate communication policies does not allow us to name the company. I am going to see if I can be a bit more precise though: "we are a company that does ...". something like this.

**Comment by vsevel** [comment]
> I like the blog @vsevel, but its a bit ambgious what/who the "we" refers to. Is there a way we can put a name on the "we" in the blog ?

@maxandersen unfortunately I am not going to be able to go more precise on this. is this still all right for a blog post?

File: _posts/2021-06-08-quarkus-deploys-apps-on-openshift.adoc (hunk 1 above)
Comment (maxandersen) [reply _posts/2021-06-08-quarkus-deploys-apps-on-openshift.adoc:46]: first hit on google when i looked up smbj - https://github.com/hierynomus/smbj

if that is not what it is even more reason to have it spelled out :)

File: _posts/2021-06-08-quarkus-deploys-apps-on-openshift.adoc (hunk 2 above)
Comment (maxandersen) [reply _posts/2021-06-08-quarkus-deploys-apps-on-openshift.adoc:57]: you are listing all the other extensions using their extension names - the others have similar rules when referred to by product name. I just suggest you use one approach not mix'em.

Comment (vsevel) [reply _posts/2021-06-08-quarkus-deploys-apps-on-openshift.adoc:57]: ah sorry. you are completely right. this is a bad replace all that I did to make sure that all "FreeMarker" had the right case. this one should not have been replaced. good catch. thanks.

File: _posts/2021-06-08-quarkus-deploys-apps-on-openshift.adoc (hunk 1 above)
Comment (vsevel) [reply _posts/2021-06-08-quarkus-deploys-apps-on-openshift.adoc:46]: that one is funny. I did not know smb was standing for "server message block". I thought it meant samba. but actually smb is the protocol, and samba a product. and smb does stand for "server message block", not samba. so yes, I will expain.

**Comment by vsevel** [comment]
hi @maxandersen I have made the different changes we discussed. let me know if that looks good to you.

**Comment by maxandersen** [comment]
@vsevel this is great! i spotted a few formatting - i'll add comments but otherwise looks good to go for me.

ok to publish it today ?

File: _posts/2021-06-08-quarkus-deploys-apps-on-openshift.adoc [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+== Enter the Quarkus Universe
+
+So less than a year ago we decided to launch a Quarkus POC on _ocpdeploy_, the goal being to deploy it as a GraalVM executable.
+The main challenges we ran into were related to the lack of support for GraalVM in some of the libraries we were using: FreeMarker, https://github.com/hierynomus/smbj[SMBJ] (a java client library that implements the Server Message Block SMB2 and SMB3 protocols), the Fabric8 OpenShift client (when instantiating OpenShift specific CRDs), and a home made java agent that ships logs to a proprietary centralized logging system.
+
+Fortunately, other bits were already there in Quarkus, such as support for command mode applications, a rest client, the Vault extension, and support for the Fabric8 Kubernetes Client, which provided a nice base for a new OpenShift Client extension that was added to the core. The Kubernetes Client extension alone was a huge push for the project because when we started we did not have Argo available, so we had to implement _apply_ and _prune_ ourselves.
+
+For FreeMarker we pushed a new extension to the Quarkiverse, with code largely inspired from work by _ppalaga_ and _carlosthe19916_.
```
Comment (maxandersen) [inline _posts/2021-06-08-quarkus-deploys-apps-on-openshift.adoc:50]:
```suggestion
For FreeMarker we pushed a new extension to the Quarkiverse, with code largely inspired from work by [ppalaga](https://github.com/ppalaga) and [carlosthe19916](https://github.com/carlosthe19916).
```

File: _posts/2021-06-08-quarkus-deploys-apps-on-openshift.adoc [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+But there are definitely some questions around sizing a build infrastructure if we had wanted to increase the number of Quarkus applications developed internally.
+
+The only drawback compared to the original solution was that our regression test suite was slower to run, essentially because our tests generated resources in the context of a specific configuration. And since we were using _MP config_, we needed to boot a new Quarkus context every time we wanted to test a different configuration. Fortunately booting a new context is extremely fast in Quarkus, but still a lot slower than with our original plain Java solution.
+
+== It is all about the community
+
+Beside work on the Vault, FreeMarker and OpenShift Client extensions, we started contributing a few PRs in and outside Quarkus, hoping to speed up the process of getting improvements, such as:
+
+- Add appendResourceVersionInObject for CRD objects https://github.com/fabric8io/kubernetes-client/pull/2365[#2365] (merged)
```
Comment (maxandersen) [inline _posts/2021-06-08-quarkus-deploys-apps-on-openshift.adoc:134]:
```suggestion
- Add `appendResourceVersionInObject` for CRD objects https://github.com/fabric8io/kubernetes-client/pull/2365[#2365] (merged)
```

**Review by maxandersen — APPROVED** [review]
LGTM :P 

add a few formatting/linking but nothing critical.

**Comment by vsevel** [comment]
thanks @maxandersen

**Pull request merged**